### ISSUE TITLE
Attempting to get snapshot by any means only yields defaultValues of atoms

### ISSUE THREAD

**Issue by nobleach** [facebookexperimental/Recoil#1523]
I'm looking for a way to dump the state of the application. The docs specify that a Snapshot is the correct way to do this. I've tried all methods: `useRecoilSnapshot`, `useRecoilCallback` and `snapshot_UNSTABLE()`. None of them seem to be able to dump my updated value. 

I've created a codesandbox to demonstrate: https://codesandbox.io/s/determined-wing-vtddn?file=/src/App.js:777-796

Overall, I'm working toward dumping the state so I can rehydrate a `<RecoilRoot />`.  I'd at least like to prove that I can dump the most up to date version of the state.

**Comment by drarmstr** [comment]
The `snapshot_UNSTABLE()` call creates a fresh snapshot that is not associated with the `<RecoilRoot>` state, so the call to `setProductId()` will never effect it.

You shouldn't call functions like `setProductId()` or `logState()` with side-effects from a React render function directly.  Something like a `useEffect()` would be more appropriate.  Note that `setProductId()` enqueues the state change, but it may not occur synchronously with the call.  (Though, it is observable immediately from subsequent updaters.)  That's why the initial snapshots still have the default value of `9`.  Though, note that the subsequent callback snapshot from `logState()` does reflect the updated value of `5`.

**Comment by nobleach** [comment]
>The snapshot_UNSTABLE() call creates a fresh snapshot that is not associated with the `<RecoilRoot>`

Ok, this makes sense. I appreciate the clarification. So If I'm understanding you correctly, Recoil has no way of dumping the current state of the application with values? 

> You shouldn't call functions like `setProductId()` or `logState()`

Absolutely agree. This was a quick codesandbox to learn how this all worked.

**Comment by drarmstr** [comment]
Yes, you should be able to get a snapshot of the current state with `useRecoilCallback()` or `useRecoilSnapshot()`.  The snapshot you got from the callback was updated, try updating the state from an effect to see if that resolves your issue with having the snapshot from `useRecoilSnapshot()` reflect the updated state as well.

**Comment by nobleach** [comment]
Yes, using `useEffect` does indeed yield the expected result. Since this code will be run during a server render, `useEffect` will not run. Since `snapshot_UNSTABLE()` runs outside of the context of the `<RecoilRoot />` I'm guessing it'll never be able to acquire a snapshot of the updated store?

Basically, I'm working to find if there's any way I can get SSR working. I realize that's out of scope of Recoil presently... I've just seen musings here and there that it may be possible with snapshots. So far I've been able to dump snapshots with default values... but there's really no path for rehydrating without taking all the data and resetting all atoms on the client side.

**Comment by drarmstr** [comment]
> Yes, using `useEffect` does indeed yield the expected result. Since this code will be run during a server render, `useEffect` will not run. Since `snapshot_UNSTABLE()` runs outside of the context of the `<RecoilRoot />` I'm guessing it'll never be able to acquire a snapshot of the updated store?
> 
> 
> 
> Basically, I'm working to find if there's any way I can get SSR working. I realize that's out of scope of Recoil presently... I've just seen musings here and there that it may be possible with snapshots. So far I've been able to dump snapshots with default values... but there's really no path for rehydrating without taking all the data and resetting all atoms on the client side.

useRecoilSnapshot() will return a Snapshot for the state of the current rendering. Recoil does support initializing state to a non-default state for the initial rendering for SSR.  You can initialize atoms using Atom Effects or the initializeState prop of <RecoilRoot>.

**Comment by nobleach** [comment]
Yeah, that `initializeState` prop takes a function that is passed a MutableSnapshot and returns void, right? I was looking at how that might work.

**Comment by drarmstr** [comment]
So you already have a `Snapshot` instance when you are trying to render?  I guess `useGotoRecoilSnapshot()` wouldn't work if you need this for an initial server rendering.  The `initializeState` prop can't just take the Snapshot as-is, since it's a mutable snapshot callback, as you point out.  Though, I suppose the interface could be updated to also accept a `Snapshot`.  Otherwise, you should be able to just manually iterate though the atoms in your snapshot and set them in the `initializeState` callback.
